examples_utils/api_tools.py

```diff
@@ -0,0 +1,50 @@
+# Copyright (c) 2023 Graphcore Ltd. All rights reserved.
+import abc
+from typing import Union, List, Any
+
+
+class Trainer(abc.ABC):
+    @abc.abstractmethod
+    def __init__(self) -> None:
+        super().__init__()
+
+    @abc.abstractmethod
+    def train(self) -> None:
+        pass
+
+    @abc.abstractmethod
+    def evaluate(self) -> None:
+        pass
+
+    @abc.abstractmethod
+    def save_hf_checkpoint(self) -> None:
+        pass
+
+
+class Pipeline(abc.ABC):
+    @abc.abstractmethod
+    def __init__(
+        self,
+        config: "ModelConfig",  # Model configuration and hyper parameters
+        hf_checkpoint: Union[str, "TransformersModel"],  # Model weights (either string or model)
```

Comment:
Why are we assuming every model checkpoint comes from Hugging Face? What if the app uses torchvision/keras/etc?